sdk/program/src/bignum.rs

```diff
@@ -0,0 +1,704 @@
+                    &z.0 as *const _ as *const u64,
+                    nonce as *const _ as *const u64,
+                    &mut bignum_ptr as *mut _ as *mut u64,
+                );
+            }
+            Self(bignum_ptr)
+        }
+    }
+
+    /// Log a `BigNum` from a program
+    pub fn log(&self) {
+        #[cfg(not(target_arch = "bpf"))]
+        crate::program_stubs::sol_log(&self.to_string());
+        #[cfg(target_arch = "bpf")]
+        {
+            unsafe { sol_log_bignum(&self.0 as *const _ as *const u64) };
+        }
+    }
+}
+
+impl fmt::Debug for BigNumber {
+    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
+        #[cfg(not(target_arch = "bpf"))]
+        {
+            use openssl::bn::BigNum;
+            let braw: &BigNum = unsafe { &*(self.0 as *mut BigNum) };
+            write!(f, "{}", braw)
+        }
+        #[cfg(target_arch = "bpf")]
+        {
+            write!(f, "{}", 0)
+        }
+    }
+}
+
+impl fmt::Display for BigNumber {
+    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
+        #[cfg(not(target_arch = "bpf"))]
+        {
+            use openssl::bn::BigNum;
+            let braw: &BigNum = unsafe { &*(self.0 as *mut BigNum) };
+            write!(f, "{}", braw)
+        }
+        #[cfg(target_arch = "bpf")]
+        {
+            write!(f, "{}", 0)
+        }
+    }
+}
+
+impl AsRef<u64> for BigNumber {
+    fn as_ref(&self) -> &u64 {
+        &self.0
+    }
+}
+
+impl From<u64> for BigNumber {
+    fn from(ptr: u64) -> Self {
+        Self(ptr)
+    }
+}
+
+/// Drop - removes the underlying BigNum
+impl Drop for BigNumber {
+    fn drop(&mut self) {
+        #[cfg(not(target_arch = "bpf"))]
+        {
+            use openssl::bn::BigNum;
+            drop(unsafe { Box::from_raw(self.0 as *mut BigNum) });
+        }
+        #[cfg(target_arch = "bpf")]
+        {
+            unsafe {
+                sol_bignum_drop(&mut self.0 as *mut _ as *mut u64);
+            }
+        }
+    }
+}
+/// Clone - creates a new BigNumber replicating itself
+impl Clone for BigNumber {
+    fn clone(&self) -> Self {
+        BigNumber::from_bytes(self.to_bytes().as_slice())
+    }
+}
+
+/// Returns an array of bytes (big endian) of self
+pub fn blake3_digest(data: &[u8]) -> Vec<u8> {
+    #[cfg(not(target_arch = "bpf"))]
+    {
+        use blake3::traits::digest::Digest;
+        blake3::Hasher::digest(data).to_vec()
+    }
+    #[cfg(target_arch = "bpf")]
+    {
+        let data_len = data.len();
+        let mut digest = [0u8; 32];
+        let mut digest_len = digest.len();
+        unsafe {
+            sol_blake3_digest(
+                data.as_ptr() as *const _ as *const u8,
+                &data_len as *const _ as *const u64,
+                digest.as_mut_ptr() as *mut _ as *mut u8,
+                &mut digest_len as *mut _ as *mut u64,
+            );
+        };
+        digest[0..digest_len].to_vec()
+    }
+}
+
+#[cfg(test)]
+mod tests {
```

Comment:
Would like to see more comprehensive testing.  Also would need testing from within a bpf program, check out the sha program in this or for an example: https://github.com/solana-labs/solana/pull/16498

Reply:
Added BPF program testing in #2b1f3e1f6ff9935bada2404e4c1c40064a8eeb71